### PR TITLE
fix: prevent planning hangs during discovery cancellation

### PR DESCRIPTION
**Pull request by anmarchenko**
## What

Prevent `ddtest plan` from hanging when full test discovery is cancelled after Test Impact Analysis returns no skippable tests.

Captured-output commands now run in their own process group on Linux and macOS so cancellation terminates framework wrappers and their descendants. A one-second `WaitDelay` also bounds output-pipe waits when a descendant has detached from that group but still holds stdout or stderr open.

## Why

`exec.CommandContext` previously killed only the direct discovery process. If `bin/rspec` or another wrapper spawned a descendant that inherited its output pipes, `CombinedOutput` continued waiting for EOF after the direct process was cancelled. Planning then remained stuck after logging `No TIA-skippable tests or suites found for this run, cancelling full test discovery`.

## E2E testing

1. In an RSpec project, use a `bin/rspec` wrapper that starts a long-lived child process inheriting stdout and stderr during `--dry-run` discovery.
2. Run `ddtest plan` with test skipping enabled for a commit where the backend returns zero skippable tests.
3. Verify DDTest cancels full discovery and exits promptly instead of waiting for the descendant to close the inherited pipes.

Automated coverage exercises normal descendants, detached descendants during cancellation, and descendants that keep output pipes open after their wrapper exits.

Validation:

- `make test`
- `make lint`

### CODE REVIEW

File: internal/ext/command_process_unix.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+//go:build darwin || linux
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/ext/command_process_unix.go]: Fixed in 4740518a. The process-group implementation now covers AIX, Android, Darwin/iOS, DragonFly, FreeBSD, illumos, Linux, NetBSD, OpenBSD, and Solaris. I also added a fallback so configureCommandProcess remains defined on js, Plan 9, and WASI. Cross-builds pass for every Unix target above, including FreeBSD/amd64; make test and make lint also pass.

[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/ext/command_process_unix.go]: Fixed in 4740518a. The process-group implementation now covers AIX, Android, Darwin/iOS, DragonFly, FreeBSD, illumos, Linux, NetBSD, OpenBSD, and Solaris. I also added a fallback so configureCommandProcess remains defined on js, Plan 9, and WASI. Cross-builds pass for every Unix target above, including FreeBSD/amd64; make test and make lint also pass.

**Comment by anmarchenko** [comment]
@autotest review

**Comment by anmarchenko** [comment]
@codex review

File: internal/ext/command_process_unix.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
+//go:build aix || android || darwin || dragonfly || freebsd || illumos || ios || linux || netbsd || openbsd || solaris
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/ext/command_process_unix.go]: Fixed in be3db44bb. Execute now installs a SIGINT/SIGTERM-cancelled Cobra context and the plan/run paths propagate it into planning. The production Test Optimization client no longer registers its independent os.Exit handler, so cancellation reaches exec.CommandContext and kills the discovery process group before DDTest exits. Planning also returns the parent context error after its goroutines finish. Context propagation/cancellation tests, make test, make lint, and a full FreeBSD/amd64 cross-build pass.

**Comment by anmarchenko** [comment]
@codex review

**Comment by anmarchenko** [comment]
@autotest review

File: internal/cmd/cmd.go [hunk 3]
```diff
@@ -165,7 +167,7 @@ func runPlanCommand(cmd *cobra.Command, args []string) {
 }
 
 func runTestCommand(cmd *cobra.Command, args []string) {
-	ctx := context.Background()
+	ctx := commandContext(cmd)
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/cmd/cmd.go]: Fixed in 0f4ec8083. Streamed test execution now has a graceful cancellation path independent of exec.CommandContext's immediate Process.Kill: the original signal is forwarded to the wrapper process group, cleanup gets 5 seconds, and timeout or a second signal escalates to a group kill. A regression test reproduces the root-context plus signal-forwarding interaction and verifies that duplicate delivery of the first signal does not skip the grace period. Deliberately detached sessions are explicitly out of scope.

[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/cmd/cmd.go]: Fixed in 0f4ec8083. DefaultCommandExecutor.Run no longer uses exec.CommandContext's immediate-kill behavior. It forwards the original cancellation signal to the wrapper process group, allows a 5-second cleanup grace period, and force-kills the group only when that period expires or a genuinely second signal arrives. The root-context and executor delivery of the same first OS signal are deduplicated. Tests cover graceful cleanup, the context/signal race, grace expiry, second-signal escalation, and ordinary descendants; deliberately detached sessions are outside the supported process tree.

**Comment by anmarchenko** [comment]
@autotest review

**Comment by anmarchenko** [comment]
@codex review

File: internal/ext/command.go [hunk 4]
```diff
@@ -97,23 +133,70 @@ func (e *DefaultCommandExecutor) Run(ctx context.Context, name string, args []st
...
+			}
+		case <-graceExpired:
+			graceExpired = nil
+			forceTermination()
 		case err := <-errChan:
 			// Command finished
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
 			return err
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/ext/command.go]: Fixed in cdc36ca7. Once cancellation is requested, Run now force-terminates any remaining ordinary descendants in the process group before returning after the wrapper exits. A regression test covers a TERM-handling wrapper with a TERM-ignoring child.

File: internal/ext/command.go [hunk 5]
```diff
@@ -97,23 +133,70 @@ func (e *DefaultCommandExecutor) Run(ctx context.Context, name string, args []st
 	notifier := e.notifier()
 	notifier.Notify(sigChan, syscall.SIGTERM, syscall.SIGINT, syscall.SIGHUP, syscall.SIGQUIT)
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/ext/command.go]: Fixed in cdc36ca7 by moving the forwarded signal list behind platform-specific helpers. The internal/ext test binary now cross-compiles successfully for js/wasm and plan9/amd64.

File: internal/ext/command_process_windows.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+func configureCommandProcess(cmd *exec.Cmd) {}
+
+func configureCommandProcessGroup(cmd *exec.Cmd) {}
+
+func signalCommand(cmd *exec.Cmd, signal os.Signal) error {
+	if cmd.Process == nil {
+		return os.ErrProcessDone
+	}
+	return cmd.Process.Signal(signal)
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/ext/command_process_windows.go]: Fixed in cdc36ca7. Windows now uses Process.Kill immediately because os.Process cannot deliver the Unix-style graceful signals, avoiding the ineffective five-second wait.

File: internal/ext/command_process_unix.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	configureCommandProcessGroup(cmd)
+	cmd.Cancel = func() error {
+		return signalCommand(cmd, os.Kill)
+	}
+}
+
+func configureCommandProcessGroup(cmd *exec.Cmd) {
+	// Start each command in its own process group so cancellation signals can
+	// terminate framework wrappers and their ordinary descendants together.
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/ext/command_process_unix.go]: This is a valid edge case for interactive test commands, but it is outside this PR's scope. The cancellation path this PR must preserve is non-interactive CI execution. Stdin remains attached because libraries such as reline inspect terminal properties even when they do not read from it; it is not currently a guarantee of prompt or debugger job-control support. Correct interactive support requires foreground process-group handoff and restoration plus coordinated job-control signal handling, which should be handled separately. We are keeping process-group cancellation unchanged here.

[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/ext/command_process_unix.go]: This is a valid edge case for interactive test commands, but it is outside this PR's scope. The cancellation path this PR must preserve is non-interactive CI execution. Stdin remains attached because libraries such as reline inspect terminal properties even when they do not read from it; it is not currently a guarantee of prompt or debugger job-control support. Correct interactive support requires foreground process-group handoff and restoration plus coordinated job-control signal handling, which should be handled separately. We are keeping process-group cancellation unchanged here.

[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/ext/command_process_unix.go]: This is a valid edge case for interactive test commands, but it is outside this PR's scope. The cancellation path this PR must preserve is non-interactive CI execution. Stdin remains attached because libraries such as reline inspect terminal properties even when they do not read from it; it is not currently a guarantee of prompt or debugger job-control support. Correct interactive support requires foreground process-group handoff and restoration plus coordinated job-control signal handling, which should be handled separately. We are keeping process-group cancellation unchanged here.

**Comment by anmarchenko** [comment]
@codex review

**Comment by anmarchenko** [comment]
@autotest review

File: internal/ext/command.go [hunk 8]
```diff
@@ -73,47 +109,100 @@ func (e *DefaultCommandExecutor) Output(ctx context.Context, name string, args [
...
+			graceExpired = nil
+			forceTermination()
 		case err := <-errChan:
 			// Command finished
+			// A wrapper can exit after handling the graceful signal while one of
+			// its ordinary descendants remains alive in the process group. Do not
+			// leave those descendants behind once cancellation has been requested.
+			if terminationRequested || ctx.Err() != nil {
+				forceTermination()
+			}
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/ext/command.go]: Thanks—the scenario is possible, but the immediate cleanup is intentional here. Once cmd.Wait reports that the wrapped command has exited during cancellation, DDTest treats that command as complete and terminates any remaining ordinary descendants so they cannot outlive the runner. The grace period applies while the wrapped command is still running; it is not a guarantee that descendants may continue after their leader exits. We accept that cancellation can truncate descendant-only report or coverage flushing in this narrow case. Preserving that work would require monitoring process-group liveness until the deadline, adding complexity and PID-reuse races without affecting normal execution or the original planning hang. For those reasons we do not consider this P1 and are keeping the current behavior.

**Comment by anmarchenko** [comment]
@autotest review

**Comment by anmarchenko** [comment]
@codex review

File: internal/testoptimization/testoptimization.go [hunk 9]
```diff
@@ -72,7 +72,9 @@ func NewTestOptimizationClientWithTelemetry(testSkippingLevel settings.TestSkipp
 	newAPITransport := func(serviceName string, level settings.TestSkippingLevel) api.Transport {
 		return api.NewTransportWithTelemetry(serviceName, level, telemetryClient)
 	}
-	client := newTestOptimizationClientWithTestSkippingLevel(nil, newAPITransport, nil, true, testSkippingLevel)
+	// The CLI owns signal cancellation so in-flight discovery commands can stop
+	// their process groups before the process exits.
+	client := newTestOptimizationClientWithTestSkippingLevel(nil, newAPITransport, nil, false, testSkippingLevel)
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/testoptimization/testoptimization.go]: this is fine

File: internal/cmd/cmd.go [hunk 10]
```diff
@@ -209,6 +219,51 @@ func runWithTelemetry(ctx context.Context, commandType telemetry.CLICommandType,
...
+}
+
+func (c commandSignalCancellation) Signal() os.Signal {
+	return c.signal
+}
+
+func commandSignalContext(parent context.Context) (context.Context, func()) {
+	ctx, cancel := context.WithCancelCause(parent)
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/cmd/cmd.go]: Fixed in c21b8909. Root cancellation signals are now platform-specific: Unix registers SIGINT, SIGTERM, SIGHUP, and SIGQUIT, while fallback targets only register supported signals. Added behavioral tests that send real SIGHUP and SIGQUIT signals and verify that the root context is canceled with the original signal as its cause, allowing CombinedOutput process-group cleanup to run.

**Comment by anmarchenko** [comment]
@codex review

File: internal/platform/python.go [hunk 11]
```diff
@@ -87,7 +87,7 @@ func (p *Python) CreateTagsMap() (map[string]string, error) {
 
 	// Execute the embedded Python script to get runtime tags
 	args := []string{"-c", pythonEnvScript, tempFile}
-	if err := p.executor.Run(context.Background(), "python", args, nil); err != nil {
+	if _, err := p.executor.CombinedOutput(ctx, "python", args, nil); err != nil {
```
[the remark this answers is not in the document]
Comment (anmarchenko) [reply internal/platform/python.go]: Fixed in 357bb871. Python, Ruby, and JavaScript runtime-tag probes now include trimmed CombinedOutput diagnostics in their returned errors while preserving the underlying error with wrapping. The existing command-failure tests now assert both the platform-specific diagnostic text and errors.Is behavior.

**Comment by anmarchenko** [comment]
@codex review

**Comment by anmarchenko** [comment]
@codex review

**Comment by anmarchenko** [comment]
## E2E Test Report: SUCCESS ✅

**Tested by:** Shepherd Agent (autonomous QA for Datadog Test Optimization)

### Test Environment

- **Method:** Local testing with the Forem RSpec playground and mockdog
- **Platform:** macOS arm64, Ruby 3.3.0, RSpec 3.12.3, Reline 0.5.3
- **Revision:** `357bb871af0c0038737926c25ad83045da1520c0`

### Results

| Check | Status |
| --- | --- |
| Reproduced full discovery with a long-lived descendant inheriting stdout/stderr | ✅ |
| Ensured the descendant was running before the empty-skippables response | ✅ |
| Empty TIA skippables cancelled full discovery | ✅ |
| Planning resumed through fast discovery and exited successfully | ✅ |
| Discovery descendant was terminated; no leaked process remained | ✅ |
| Plan contained the expected test file and runner split | ✅ |
| Full `ddtest run` executed a Reline EOF spec | ✅ — 1 example, 0 failures |
| `ddtest` unit tests and lint | ✅ |
| PR CI at the tested revision | ✅ |

### Test Methodology

1. Injected the PR branch into Forem with Crook and enabled debug logs.
2. Temporarily augmented `bin/rspec --dry-run` to spawn a long-lived child that inherited stdout and stderr, recording its PID.
3. Used the `ddtest-empty-skippables` mockdog scenario and delayed only the skippables response by one second so the child was definitely alive before cancellation.
4. Verified the cancellation log, fast-discovery fallback, exit code 0, generated plan artifacts, and that the recorded child PID no longer existed.
5. Ran a full one-file `ddtest run` whose RSpec example called `Reline.readline`; it received EOF and completed successfully.

### Issues Found

No product issues were found in the requested discovery-cancellation and Reline scope.

### Verification

This process-lifecycle change does not alter backend request or event shapes, so Datadog UI verification was not applicable. The mockdog-backed full run recorded one passing test, suite, module, and session with valid telemetry.

---

_This E2E test was performed by Shepherd — autonomous QA for Datadog Test Optimization._

**Review by gnufede — APPROVED** [review]
